File: day4/puzzle2.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> read_input(){
    ifstream fin("input.txt");
    vector<string> res;
    string s;
    while (getline(fin, s)){
        res.push_back(s);
    }
    return res;
}

int solve(const vector<string>& matrix){
    int n = matrix.size();
    int m = matrix[0].size();
    int res = 0;
    for(int i = 1; i < n - 1; ++i){
        for(int j = 1; j < m - 1; ++j){
            if (matrix[i][j] == 'A'){
                if (((matrix[i - 1][j - 1] == 'M' && matrix[i + 1][j + 1] == 'S') || (matrix[i - 1][j - 1] == 'S' && matrix[i + 1][j + 1] == 'M'))
                    && ((matrix[i - 1][j + 1] == 'M' && matrix[i + 1][j - 1] == 'S') || (matrix[i - 1][j + 1] == 'S' && matrix[i + 1][j - 1] == 'M'))){
                    res++;
                }
            }
        }
    }
    return res;
}

int main(){
    vector<string> matrix = read_input();
    cout << solve(matrix) << endl;
    return 0;
}

File: day13/puzzle2.py
from sympy import symbols, Eq
import sympy
import regex as re
def read_input(str):
    with open(str, "r") as f:
        pattern = r"""
            Button\sA:\sX\+(\d+),\sY\+(\d+)\n
            Button\sB:\sX\+(\d+),\sY\+(\d+)\n
            Prize:\sX=(\d+),\sY=(\d+)
        """
        lines = f.read()
        matches = re.findall(pattern, lines, re.VERBOSE)
        results = [
        {
            "a": int(match[0]),
            "b": int(match[2]),
            "c": 10000000000000 + int(match[4]),
            "d": int(match[1]),
            "e": int(match[3]),
            "f": 10000000000000 + int(match[5])
        }
        for match in matches
    ]

    return results

def solve(equations):
    res = 0
    for eq in equations:
        #print(eq)
        a, b, c, d, e, f = eq["a"], eq["b"], eq["c"], eq["d"], eq["e"], eq["f"]
        numerator = d * c - a * f
        denominator = d * b - a * e
        if denominator == 0:
            continue
        if numerator % denominator != 0:
            continue
        y = numerator // denominator
        if y < 0:
            continue
        #print(y)
        numerator = c - b * y
        denominator = a
        if denominator == 0:
            continue
        if numerator % denominator != 0:
            continue
        x = numerator // denominator
        #print(x)
        if x < 0:
            continue
        res += 3 * x + y
    return res

if __name__ == "__main__":
    equations = read_input("input.txt")
    print(solve(equations))